### DISCUSSION TITLE
Adding a new variable to the system memory

### DISCUSSION THREAD

**Discussion by mmsor87** [GazzolaLab/PyElastica#351] [Q&A] [answered]
Hi,

I'm creating a program that is modelling slender structures in the ocean based on your library. In this process, I've created a class in 'interactions.py' based on Morison's equation to include inertial effects in the flow-structure interaction between the rod and its surrounding fluid.

The apply forces function in my class looks like this

```
def apply_forces(self, system, time):
        
        wave_velocity = self.AiryWaveTheory(system.lengths,
                                    system.position_collection,
                                    time)
        
        current_velocity = self.Current(system.lengths)
        
        fluid_velocity = wave_velocity + current_velocity
        
        morison_force = self.compute_morison_forces(system.velocity_collection,
                                               system.acceleration_collection,
                                               system.lengths,
                                               system.radius,
                                               system.mass,
                                               fluid_velocity)
        
        elements_to_nodes_inplace(morison_force, system.external_forces)
```

where the wave-induced inertia is applied through the airy wave theory and kinematic stretching, while the current velocity is a distributed force decreasing linearly from the sea level to the seabed. The "compute_morison_forces" function applies the Morison forces to each element similarly to the "Slender Body Theory" function (https://docs.cosseratrods.org/en/latest/api/external_forces.html#elastica.interaction.SlenderBodyTheory). 

The class is working as intended and has been validated against OrcaFlex in its current state, but I still need to include the fluid acceleration for the class to be complete. This means that I must save the fluid velocity in the "system" object to find the fluid acceleration at the next step. 

Being a novice Python programmer, I'd really appreciate some input to where in the code I should add this functionality (It looks to me like the "memory_block\memory_block_rod.py" is the correct place, and that the best approach would be to store the fluid velocity at nodes similarly to the "position_collection" in the "allocate_block_variables_in_nodes" function), or if there is another simpler approach to save the "fluid velocity" parameter from my apply forces function for one time step to calculate the acceleration?

**Comment by armantekinalp** [comment]
@sy-cui can you help this issue.

**Comment by sy-cui** [comment, marked as answer]
Hi,

You are correct that `elastica/memory_block/memory_block_rod.py` is the place to add this variable internally. Assuming that you want to add a vector variable called `fluid_velocity` on the nodes of a rod, there are three files you want to change:
1. `elastica/memory_block/memory_block_rod.py`. In the `_allocate_block_variables_in_nodes` method you mentioned, simply add to the `map_vector_dofs_in_rod_nods` dictionary the line `"fluid_velocity": 3`. The block initialization will automatically go through each item in the dictionary and allocate the memory as vectors stored in rod nodes. Of course if you decide to put them on rod elements instead, you will need to put this in the `_allocate_block_variables_in_elements` method. 
2. `elastica/rod/Cosserat_rod.py`. Strictly speaking, in the memory_block file we are not initializing these variables, but moving them into a contiguous block of memory. The actual initialization interface is provided in the `Cosserat_rod.py` file instead. In the `CosseratRod` class, you need to put `fluid_velocity` in the [`__init__()`](https://github.com/sy-cui/PyElastica/blob/d68887cb72649c8b24227d57f68bca93850a7bf6/elastica/rod/cosserat_rod.py#L148) function in the same way as the other variables. This will enable the usage of `system.fluid_velocity`. Then in the `straight_rod()` and `ring_rod()` method, you may need to put them in the same list whenever you see the full list of variables, because you need to pass the new variable from the return list of `allocate()` to `self.fluid_velocity`. 
3. `elastica/rod/factory_function.py`. This function is called within the `Cosserat_rod.py` file. To add `fluid_velocity`, do it the same way as `external_forces`, for example. Just allocate the memory and return it.

Note that this is not the only way to make this work. You may also consider allocating this variable outside the main memory block, like in your interaction class. However, you will need to know the exact size of this `fluid_velocity`, which is not difficult. For example, consider the following:
```py
class MorrisonForce(NoForces):
    def __init__(self, fluid_velocity: np.ndarray):
        # Create an aliase to fluid_velocity
        self.fluid_velocity = fluid_velocity
    
    def apply_forces(self, system, time=0.0):
        self.compute_morrison_force(..., self.fluid_velocity, ...)
```
and in the main case file,
```py
...
# Initialize fluid velocity as a on-node buffer
fluid_velocity = np.zeros((3, n_elements + 1), dtype=np.float64)
simulator.add_forcing_to(rod).using(MorrisonForce, fluid_velocity=fluid_velocity.view()) # .view() is optional

# use fluid_velocity to compute acceleration somehow
```

Hope this helps.

**Reply by mmsor87** [reply]
Thank you very much for a really helpful reply, adding it to the main memory block did the trick - I've added both fluid_velocity and fluid_acceleration to the system.
